apps/pkt-gen/pkt-gen.c

```diff
@@ -1177,6 +1246,74 @@ send_packets(struct netmap_ring *ring, struct pkt *pkt, void *frame,
 	return (sent);
 }
 
+/*
+ * Send two ARP packets: a gratuitous ARP to advertise the netmap device, and
+ * an ARP request directed at the remote party.
+ *
+ * Many switches require regular ARP traffic to keep the [mac:ip] association
+ * to the physical switch port.  When the association is lost the switch will
+ * broadcast your pkt-gen traffic to *all* ports, killng network performance.
+ * This often happens during PHY reset, when switching into netmap mode.
+ */
+static void
+send_arp(struct targ *targ)
+{
+	struct nm_desc *nmd = targ->nmd;
+	struct netmap_if *nifp = targ->nmd->nifp;
+	struct ether_addr *src_mac = &targ->g->src_mac.start;
+	uint32_t src_ip = htonl(targ->g->src_ip.ipv4.start);
+	uint32_t dst_ip = htonl(targ->g->dst_ip.ipv4.start);
+	int vhlen = targ->g->virt_header;
+	int need = 2;
+	if (!targ->garp) return;
+
+	for (int i = nmd->first_tx_ring; i <= nmd->last_tx_ring && need > 0; i++) {
+		void *frame;
+		struct pkt *pkt;
+		struct arphdr *hdr;
+		struct netmap_slot *slot;
+		struct netmap_ring *ring = NETMAP_TXRING(nifp, i);
+
+		if (nm_ring_empty(ring))
+			continue;
+
+		slot = &ring->slot[ring->cur];
+		frame = NETMAP_BUF(ring, slot->buf_idx);
+		pkt = (struct pkt*)(frame - sizeof(pkt->vh) + vhlen);
+		hdr = &pkt->arp4.ea_hdr;
+		slot->len = ARP_SIZE(vhlen);
+		slot->ptr = 0;
+		bzero(frame, vhlen);
+
+		pkt->eh.ether_type = htons(ETHERTYPE_ARP);
+		memset(pkt->eh.ether_dhost, 0xff, ETH_ALEN);
+		bcopy(src_mac, pkt->eh.ether_shost, ETH_ALEN);
+		hdr->ar_hrd = htons(ARPHRD_ETHER);
+		hdr->ar_pro = htons(ETHERTYPE_IP);
+		hdr->ar_hln = ETH_ALEN;
+		hdr->ar_pln = IP4_ALEN;
+
+		if (need == 2) {	/* first send a GARP */
+			hdr->ar_op = htons(ARPOP_REPLY);
+			bcopy(src_mac, pkt->arp4.arp_sha, ETH_ALEN);
+			bcopy(src_mac, pkt->arp4.arp_tha, ETH_ALEN);
+			bcopy(&src_ip, pkt->arp4.arp_spa, IP4_ALEN);
+			bcopy(&src_ip, pkt->arp4.arp_spa, IP4_ALEN);
```

Comment:
The two last lines are the same... did you mean to use `dst_ip` in the second one?